### PR TITLE
lantiq/xway: add support for FritzBox7312

### PR DESCRIPTION
**Pull request by StefanFreifunk**
https://wiki.openwrt.org/toh/avm/fritz.box.wlan.7312

Specification:
- SoC: Lantiq Xway ARX188 PSB 50812 EL (MIPS 34Kc), 393MHz
- RAM: DDR1 64MiB @ 196 MHz
- Storage: Macronix MX29GL128EHT2I-90G, 16MiB
- Wireless: internal Atheros AR9227 b/g/n 2x2
- Ethernet: 1x100M
- xDSL: lantiq ADSL2+

Installation (FTP):
set host IP to 192.168.178.3, power on FritzBox, immediately start FTP connection (did not work? - try again connecting a bit later), first 5seconds it accepts ftp connections - ping 192.168.178.1 can help
- ftp 192.168.178.1
- User: adam2
- Pwd: adam2
- bin
- passiv
- quote MEDIA FLSH
- put firmware.image mtd1
- quote REBOOT
- quit

### CODE REVIEW

**Comment by hauke** [comment]
Could you please squash all commits into and add a Signed-of-by line to the commit please.
Otherwise it looks good to me.

File: target/linux/lantiq/dts/FRITZ7312.dts [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	chosen {
+		bootargs = "console=ttyLTQ0,115200";
+	};
+
+	aliases {
+		led-boot = &power;
+		led-failsafe = &power;
+		led-running = &power;
+
+		led-internet = &info_green;
```
Comment (mkresin) [inline target/linux/lantiq/dts/FRITZ7312.dts]: please drop led-internet here. You do not have a distinct led to show the internet connection status. The use of a single led for dsl line status  and internet connection status is possible since 0481a2a2c5b85e82d3b7a55ee556c1a23dd1c135 and all affected boards were changed with 259fc1e77892ef8b7fb41c5bc118b55e31e0554b.

File: target/linux/lantiq/dts/FRITZ7312.dts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+		led-dsl = &power;
```
Comment (mkresin) [inline target/linux/lantiq/dts/FRITZ7312.dts]: please use info green as led-dsl.

File: target/linux/lantiq/dts/FRITZ7312.dts [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+		};
+		dect {
+			label = "fritz7320:green:dect";
+			gpios = <&gpio 38 GPIO_ACTIVE_LOW>;
+		};
+		wlan: wlan {
+			label = "fritz7320:green:wlan";
+			gpios = <&gpio 37 GPIO_ACTIVE_LOW>;
+		};
+		info_green: info_green {
```
Comment (mkresin) [inline target/linux/lantiq/dts/FRITZ7312.dts]: According to the manual, there should be an info red led as well. Would you please add this one as well.

File: target/linux/lantiq/dts/FRITZ7312.dts [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+			label = "dect";
+			gpios = <&gpio 2 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_PHONE>;
+		};
+	};
+
+	gpio-leds {
+		compatible = "gpio-leds";
+		power: power {
+			label = "fritz7320:green:power";
```
Comment (KanjiMonster) [inline target/linux/lantiq/dts/FRITZ7312.dts]: Your file says fritz7315, but your leds say fritz7320 - copy-paste error?

Comment (StefanFreifunk) [reply target/linux/lantiq/dts/FRITZ7312.dts]: Yes, copy-paste xP

File: target/linux/lantiq/dts/FRITZ7312.dts (hunk 3 above)
Comment (StefanFreifunk) [reply target/linux/lantiq/dts/FRITZ7312.dts]: I'll test leds. On device board it didn't see a 2nd led @info - yay, manual says there is a red one.

**Comment by oaaeruak** [comment]
Thanks for your work! Would be nice to see it merged upstream :)

File: target/linux/lantiq/dts/FRITZ7312.dts (hunk 3 above)
Comment (StefanFreifunk) [reply target/linux/lantiq/dts/FRITZ7312.dts]: tried gpios 24 to 56 in /source/target/linux/lantiq/dts/FRITZ7312.dts using LuCi after compile, aka:

		info_red {
			label = "fritz7312:red:info";
			gpios = <&gpio 24 GPIO_ACTIVE_LOW>;
		};
		info_red2 {
			label = "fritz7312:red:info";
			gpios = <&gpio 25 GPIO_ACTIVE_LOW>;
		};
	...
		info_redX {
			label = "fritz7312:red:info";
			gpios = <&gpio 56 GPIO_ACTIVE_LOW>;
		};

no red led could be activated so far.

if -> gpio 34 GPIO_ACTIVE_LOW -> connection loss an Lan1 (has only one)
if -> gpio 24 GPIO_ACTIVE_LOW -> bootloop
if -> 35 GPIO_ACTIVE_HIGH -> led "broken"
if -> 54+55+56 GPIO_ACTIVE_HIGH -> led broken

Comment (mkresin) [reply target/linux/lantiq/dts/FRITZ7312.dts]: I neither wasn't able to find the GPIO of the mentioned red info led in the sources. But I found references to some of the mentioned GPIOs:

- GPIO#34: ext phy1 reset
- GPIO#54: dect reset

Seem to me the manual is wrong and there isn't a red info led.

Comment (StefanFreifunk) [reply target/linux/lantiq/dts/FRITZ7312.dts]: ill take "info_red" out for now

**Comment by StefanFreifunk** [comment]
Hope commit is as expected now - first time doing this

**Comment by Redfoxmoon3** [comment]
The commits aren't squashed yet, here this might help: https://lede-project.org/submitting-patches#squashing_commits

**Comment by nazar-pc** [comment]
Commits can be squashed on merge in GitHub UI if you use it for this

**Comment by mkresin** [comment]
I've merged your PR to my [staging tree](https://git.lede-project.org/?p=lede/mkresin/staging.git;a=summary). I've done the following changes:

- add back led-failsafe and led-running to the dts (they are indeed required and my former advice to remove them was rubbish; sorry about that)
- squashed the commits into one
- changed the led names from fritz7320 to fritz7312 in the dts
- removed the green:power led from 01_leds; it isn't required with correct led mapping in the dts (and shouldn't work since the name differs from the one used in the dts [fritz7312 vs. fritz7320]
- removed the green:wlan led from 01_leds; it shouldn't be required due to the led mapping in the dts + ath9k driver + the default logic at the top of the file (beside of that, same name mismatch as for the power led)
- changed the author of the commit(s) to the name + mail address used in the signed off by line (please fix your config!)

Please build an image from my staging tree, test if everything is still working as expected and report back here.

**Comment by StefanFreifunk** [comment]
Thanks, i'll test mkresin's image.

**Comment by xdarklight** [comment]
@StefanFreifunk could you please also test if ethernet is working for you?

the current ethernet driver in LEDE uses a 2.5MHz MDC clock.
However, the AVM kernel (source-files-FRITZ.Box_7312.06.01/linux-2.6.32/drivers/net/avm_cpmac/switch/ifx/ar9/swi_ar9.c in avmnet_swi_ar9_hw_setup) sets the MDC clock to 1MHz.

I made some tests with my own 7312 a while ago and without changing the MDC clock ethernet was not working (I can't remember though if the PHY wasn't detected or if it was but simply no traffic was flowing). if you want to have a look at the four patches I wrote for OpenWrt back in the days then you can have a look at this branch: https://github.com/xdarklight/openwrt/commits/fb7312_wip_20160114

**Comment by StefanFreifunk** [comment]
working	:D
Bootlog: https://wiki.openwrt.org/toh/avm/fritz.box.wlan.7312#lede_snapshot

-> WLanButton+WLan workin

-> got 1st Boot Lan bug after flashing	> no DHCP-ip	> DHCP-ip after reboot	> Lan workin, see Ethernet: https://wiki.openwrt.org/toh/avm/fritz.box.wlan.7312#status

@xdarklight see: https://wiki.openwrt.org/toh/avm/fritz.box.wlan.7312#firmware_build_fixes
-> got this by comparing .dts and other files + trial&error .)
ucidef_set_interface_lan 'eth0' → prevents LanPort not responding/link loss
set pinmux to default & remove custom → prevents link loss

-> LEDs workin out of Box
-> LuCI workin

-> did not test ADSL again

BUGs&Features:
-> @end_of_bootlog: daemon.err uhttpd[807]: sh: write error: Broken pipe -> hmm, no disadvantages so far

**Comment by StefanFreifunk** [comment]
After some testing found a BUG:

-> tried to set eth0 (single lan port) as WAN uplink (dhcp-client mode) -> eth0 is up but gets no IP-address (got one or 2 cases it got an IP after playing around for a while, could not reproduce it)

-> if eth0 (lan port) is bridged to wlan (stock lede configuration: static address, dhcp-server running) eth0 is up and IPs are assigned to clients both on lan and wlan

File: target/linux/lantiq/dts/FRITZ7312.dts (hunk 3 above)
Comment (StefanFreifunk) [inline target/linux/lantiq/dts/FRITZ7312.dts]: After some testing found a BUG?

-> tried to set eth0 (single lan port) as WAN uplink (dhcp-client mode) -> eth0 is up but gets no IP-address (got one or 2 cases it got an IP after playing around for a while, could not reproduce it, no internet connectivity)

common case:
-> if eth0 (lan port) is bridged to wlan (stock lede configuration: static address, dhcp-server running) eth0 is up and IPs are assigned to clients both on lan and wlan

-> have both Lan&Wan to be bridged?
- wifi is working if put from bridged LANinterface to an other interface serving as dhcp-server,
- eth0 (single LanPort) gets no ip in dhcp-client mode (both in LANinterface or another f.e. WAN),
**in syslog i can see uhcpd sending dhcp discoveries, but gets no lease** -> connected to an other router (dhcpsever) or cable modem (dhcpserver)
- also tried vlan tagging aka: eth0.1 (port1 lan) and eth0.2 (port 2 wifi)

-> swconfig shows no switch, is there an internal ar9 one?

-> also played around with commits from https://github.com/xdarklight/openwrt/commits/fb7312_wip_20160114, comparing and adding them to https://git.lede-project.org/?p=lede/mkresin/staging.git;a=summary
- did not change behavior - would leave them out, "less is more"

-> think its somewhere in software: dhcp? routing?

**Comment by xdarklight** [comment]
Hi Stefan,

sorry for the huge delay, I wanted to test this on my 7312 weeks ago.

you wrote "set pinmux to default & remove custom → prevents link loss"
could you please be more precise and explain which part prevents link loss (and even include this in the commit message)?
my version back then (as you noticed: with more patches) had pinmux settings in the .dts and it was working fine for me: https://github.com/xdarklight/openwrt/commit/b63e57fca57f7f00815d21feac6519fd564532d1#diff-1d7cf20937fd23270ff53e233bd08562R31

does your 7312 also have an AT8030 Ethernet PHY (maybe you could update the commit message to include the PHY as well)?
this PHY has a known hardware bug: https://elixir.free-electrons.com/linux/v4.14/source/drivers/net/phy/at803x.c#L334
GPIO 34 is connected to the reset line of this PHY (at least on my board).
these steps should allow the PHY driver to work around this PHY hardware bug (which this might also explain the sporadic network connectivity issues which you have seen on your device):
* enabling CONFIG_AT803X_PHY in the kernel (which is currently disabled for the "xway" subtarget)
* and adding a sub-node for the PHY along with the reset-gpios property
  * this however requires something like https://github.com/xdarklight/openwrt/commit/0c51eb25e1a8f61e0e981362850545791e6b1ed0 first

regarding your AR9 switch question: it's complicated :)
the AR9 has a 3-port switch (one port is reserved and connected to the CPU).
however, the current driver only implements one port and does not expose it as switch

I cannot say much about the bridge use-case as I run relayd myself to create a pseudo-bridge.

back to the actual topic: testing your patch on my own 7312
I used this version from Mathias' tree: https://git.lede-project.org/?p=lede/mkresin/staging.git;a=commitdiff;h=ea8cd26198ba0b8e29b07e9d77b4051ef470e4b0
* all LEDs are working fine
  * (I like that you used GPIO_ACTIVE_LOW / GPIO_ACTIVE_HIGH - unlike I did in my patch)
* wifi and Ethernet MAC addresses are set (although I haven't compared them with the stock firmware)
  * (also better than in my patch)
* wifi seems to work fine
* Ethernet also seems to work fine at first sight
  * this might change though when resetting the Ethernet PHY (I think this is when I started my etop patch - but I can't remember exactly anymore)!

-----

to sum it up:
* I think your patch looks pretty good in general
* due to the hardware bug in the PHY the driver should be enabled and configured (otherwise 7312 support might break as soon someone enables CONFIG_AT803X_PHY in the xway kernel config)
  * I think AVM also does this: ``source-files-FRITZ.Box_7312.06.01/linux-2.6.32/drivers/net/avm_cpmac/configs/config_7312.h`` -> AVMNET_CONFIG_FLAG_RESET, which then results in a call to phy_reset in ``source-files-FRITZ.Box_7312.06.01/linux-2.6.32/drivers/net/avm_cpmac/phy/avmnet_ar803x.c``
* adding the correct pinmux settings is always a good idea to prevent users from playing with these "reserved" pins through /sys/class/gpio/... (this however only works if the pinmux is configured correctly)


Regards
Martin

**Comment by mkresin** [comment]
Hey Stefan,

since I haven't merged your PR to master yet, I'll reopen it.

The AT8030 Ethernet PHY should be mentioned in the commit message, the driver should be enabled in the kernel config and the reset GPIO should be set in the dts.

We had the issue described by Martin with the [AVM Fritz!Box 7360SL](https://bugs.lede-project.org/index.php?do=details&task_id=343) and it was fixed by adding the reset gpios with 7dd42d610dcd6f9271d5d44edff52caf633b8abf.

**Comment by mkresin** [comment]
Please reopen the PR if you have the time to address he requested changes.

**Comment by StefanFreifunk** [comment]
Sorry for delay, tried mkresin's & xdarklight's hints, still no luck. I'll look if there's something in FritzBox7312 sources.
